designs/2020-generic-ast-support/README.md

```diff
@@ -0,0 +1,57 @@
+- Repo: eslint/eslint
+- Start Date: 2020-06-04
+- RFC PR: 
+- Authors: Ilya Volodin
+
+# Generic AST support
+
+## Summary
+
+Allow ESLint to traverse and lint generic (non estree compliant AST) and allow parsers to specify information about AST that would drive ESLint's traversal and features.
+
+## Motivation
+
+ESLint has cemented itself as a go to tool for JavaScript/TypeScript ecosystem for style checking and linting. However, there are a lot of connected languages/technologies that are used every day by ESLint users that are either lacking linters, or require yet another tool to be setup and configured for the repository. Examples include HTML, CSS, SCSS, LESS, GraphQL, JSON, etc.
+With addition of config overrides users can now setup separate configuration per glob, so that would allow them to have a single instance of ESLint and a single configuration that would allow linting of the entire repository, including all non javascript/typescript files.
+
+## Detailed Design
+
+Currently ESLint has a lot of hardcoded assumptions about AST. For example, it assumes that nodes are identified by the property called `type` and that property exists on every node. It assumes that top-most node is always `Program` and that AST supports code-path-analysis (which is very specific to JavaScript). This RFC proposes to modify expected output from the `parseForESLint` method of the custom parsers to include additional property called `parserSupportOptions` of type object with four new properties (for now):
+ - nodeIdentifier: string - Specifies name of the ASTNode property that identifies node. Defaults to `type` for backwards compatibility.
+ - supportCodePathAnalysis: Boolean - Indicates AST's support for built-in CodePathAnalysis support. Defaults to `true` for backwards compatibility.
+ - supportScope: Boolean - Indicates AST's support for scope analysis. Defaults to `true` for backwards compatibility.
```

Comment:
Custom parsers can already return a custom `scopeManager` property. When I was playing around with another non-ESTree AST, I threw together a do-nothing [`ScopeManager`](https://eslint.org/docs/developer-guide/scope-manager-interface) implementation. Similar to the above idea, what if we add `scopeManager: null` as a valid value that serves the same purpose as `supportScope: false` would here?

Reply:
I added this property because ESLint's documentation is incorrect. Right now it states that you can return `null` from parser for `scopeManager`, however, that's not true. ESLint will accept `null`/`undefined` `scopeManager`, but will blow up, because it will try to navigate it. Because of that, I had create `scopeManager` like this:
```js
scopeManager: {
    variables: [], scopes: [{ set: new Map(), variables: [], through: [] }], globalScope: { set: new Map(), variables: [] }, getDeclaredVariables: () => {}
}
```
And then there's a bunch of code in the core that tries to analyze and attach those scopes to nodes. But I think if we can accept `null` as `scopeManager` and disable core attachment code, that would be perfectly fine from my side.

Reply:
Since `scopeManager` and `codePathAnalyzer` could often be used together, I think we should make them consistent. That would argue for either 1) making `scopeManager` accept `boolean | ScopeManager` or 2) using the existing `scopeManager` documentation as precedent and fixing it to allow `null` and having `codePathAnalyzer` accept `null`. I agree with your argument for case (1) in the `codePathAnalyzer` discussion above, and if `scopeManager: null` doesn't work anyway, changing it to `boolean | ScopeManager` wouldn't break anything that isn't already broken.

Reply:
@nzakas What if you do want to use existing one (i.e. eslint's built-in codePathAnalyzer)? How do you specify that with null | object? I would imagine that both babel and TypeScript parser would want to continue using built-in codePathAnalyzer for some time, until the have a need to create their own implementation.

Reply:
I'd envision that simply not specifying `scopeManager` or `codePathAnalyzer` would mean "use the default".

Reply:
So it's a tri-state? `object|null|undefined`? Ok, I can work with that, although to be honest, to me, `boolean|object` seems a lot more descriptive.

Reply:
I see how `object | null | undefined` is the most literal way to type the option, but I also see how `object | boolean` is friendlier. So I'll throw out a bad idea for the purposes of discussion:

If I were doing this in a language that had a [variant type](https://reasonml.github.io/docs/en/variant#constructor-arguments) I might do something like this:

```reason
type CodePathAnalyzerConfig =
    | Off     /* Or None, Disabled */
    | Default /* Or On, BuiltIn */
    | Custom(CodePathAnalyzer);
```

Since we can't quite do that with JS, what about this?

```js
type ParseForESLintReturnValue = {
    // ...
    codePathAnalyzer?:  // Optional
        | "off"     // Or "none", "disabled"
        | "default" // Or "on", "builtin"
        | CodePathAnalyzer,
}
```

If the property is missing/`undefined`, that's the same as `"default"`.

Reply:
I like this a bit better then `null|undefined|function`, it's a bit clearer in it's intentions. Although I still think `boolean|function` is easier to understand. But difference is minor.

Reply:
Works for me